workshops/util.py

```diff
@@ -877,7 +874,25 @@ def merge_objects(object_a, object_b, easy_fields, difficult_fields,
 
 
 def only_for_admins(f):
-    f = user_passes_test(lambda user: user is not None and user.is_admin)(f)
+    assert not hasattr(f, 'access'), \
+        'You already used some access control decorator on this view.'
+    f = user_passes_test(is_admin)(f)
+    f.access = 'only_for_admins'
+    return f
+
+
+def only_for_logged_in(f):
+    assert not hasattr(f, 'access'), \
+        'You already used some access control decorator on this view.'
+    f = login_required(f)
+    f.access = 'only_for_logged_in'
+    return f
+
+
+def login_not_required(f):
+    assert not hasattr(f, 'access'), \
+        'You already used some access control decorator on this view.'
+    f.access = 'login_not_required'
```

Comment:
Okay, I understand why you're setting `f.access` here.
1. docstrings!!!
2. functions and classes can set their own parameters, and you can overwrite them. I'd use `._access_control` or some other "hidden" name instead of `.access`.

What's more is that this kind of decorators basically hide "permissions by inheritance" model @gvwilson was talking about yesterday.

Writing `@perms_required(['workshops.do_a', 'workshops.do_b'])` gives us more control over the perms (and I know at least one place where we have to use this approach to give only specific parts accessible to some user groups), while writing `@groups_required(['administrators'])` is cozy and can make me lazy (`@only_for_admins` and `@only_for_logged_in` are basically other names for decorators that already exist) -- I can forget to update `administrators` permissions list, for example.

I kinda feel mixing different decorators is not good.

Finally there's some naming convention I feel would be better here. I like `_required` approach, so maybe:

* `admin_required` instead of `only_for_admins`
* Django's `login_required` imported as `django_login_required` and `only_for_logged_in` renamed to `login_required`
* `login_not_required` is fine

@gvwilson any additional comments? Especially regarding more control (explicit permissions list) vs less work (explicit/implicit group-based permissions).